mesonbuild/modules/fs.py

```diff
@@ -39,13 +39,15 @@ def __init__(self, interpreter: 'Interpreter') -> None:
         super().__init__(interpreter)
         self.snippets.add('generate_dub_file')
 
-    def _absolute_dir(self, state: 'ModuleState', arg: str) -> Path:
+    def _absolute_dir(self, state: 'ModuleState', arg: 'FileOrString') -> Path:
```

Comment:
you need to have a `FeatureNew()` check for files as well, it needs to be on each entry point so the name printed is correct, not on the helper.

Reply:
I am sorry but I do not fully understand what you mean this may be caused by my lack of knowledge of the meson code base.

So please correct me if I am wrong!

You would like a `@FeatureNew` check for each fs method newly accepting a File argument?

Something like:

```python
@noKwargs
@typed_pos_args('fs.is_symlink_file', File)
@FeatureNew('fs.is_symlink_file', '0.58.0')
def is_symlink_file(self, state: 'ModuleState', args: T.Tuple[File], kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
    p = Path(arg.absolute_path(state.source_root, self.interpreter.environment.get_build_dir()))
    return ModuleReturnValue(p.is_symlink(), [])

@noKwargs
@typed_pos_args('fs.is_symlink_str', str)
def is_symlink_str(self, state: 'ModuleState', args: T.Tuple[str], kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
    return ModuleReturnValue(self._absolute_dir(state, args[0]).is_symlink(), [])

@noKwargs
@typed_pos_args('fs.is_symlink', (str, File))
def is_symlink(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
    if isistance(args[0], File):
        return self.is_symlink_file(args[0])
    return self.is_symlink_str(args[0])
```

Without modifying `_absolute_dir` at all ?

But this seems really ugly and it would introduce new user visible functions.
Is there a way easier to use the `@FeatureNew` decorator for only a subset of accepted argument types?

And how I understand the code the helper has no check at all except the mypy hints and relies
on the `@typed_pos_args` checks before each entry function, which we have.

Reply:
You don't need to make 2 separate methods:
```meson
@noKwargs
@typed_pos_args('fs.is_symlink', (str, File))
def is_symlink(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
    if isistance(args[0], File):
        FeatureNew('fs.is_symlink_file', '0.58.0').use(state.subproject)
    return self._absolute_dir(state, args[0]).is_symlink()
```

Reply:
Thanks @xclaesse.
My python foo was not good enough to fully understand the FeatureNew code.
I thought of it only as a function wrapper.
Now there are FeatureNew checks for each user facing function taking a File object.